src/Application/Game.cpp

```diff
@@ -1418,103 +1418,80 @@ void Game::EventLoop() {
-                                    pParty
-                                        ->pPlayers[(unsigned __int8)
-                                                       town_portal_caster_id]
-                                        .CanCastSpell(0x14u);
-
-                                    pMessageQueue_50CBD0->AddGUIMessage(
-                                        UIMSG_Escape, 1, 0);
-                                    continue;
-                                }
-                                v63 = 210;
-                                break;
-                        }
-                    } else {
-                        v63 = 206;
+                    //if (uGameState == GAME_STATE_CHANGE_LOCATION) continue;
+                    switch (uMessageParam) {
+                        case 0:
+                            v63 = 206;
+                            break;
+                        case 1:
+                            v63 = 208;
+                            break;
+                        case 2:
+                            v63 = 207;
+                            break;
+                        case 3:
+                            v63 = 211;
+                            break;
+                        case 4:
+                            v63 = 209;
+                            break;
+                        case 5:
+                            v63 = 210;
+                            break;
+                        default:
+                            Assert(true && "Bad TP param");
+                            break;
                     }
-                    if (!_449B57_test_bit(pParty->_quest_bits, v63)
-                        && !engine->config->debug.TownPortal.Get())
-                        return;
-                    goto LABEL_486;
+
+                    // check if tp location is unlocked
+                    if (!_449B57_test_bit(pParty->_quest_bits, v63) && !engine->config->debug.TownPortal.Get())
+                        continue;
+
+                    // begin TP
+                    SaveGame(1, 0);
+                    v64 = pMapStats->GetMapInfo(pCurrentMapName);
+                    v65 = uMessageParam;
+                    // if in current map
+                    if (v64 == TownPortalList[uMessageParam].uMapInfoID) {
+                        pParty->vPosition.x = TownPortalList[v65].pos.x;
+                        pParty->vPosition.y = TownPortalList[v65].pos.y;
+                        pParty->vPosition.z = TownPortalList[v65].pos.z;
+                        pParty->uFallStartZ = pParty->vPosition.z;
+                        pParty->sRotationZ = TownPortalList[v65].rot_y;
+                        pParty->sRotationY = TownPortalList[v65].rot_x;
+                        // take mana
+                        pParty->pPlayers[(unsigned __int8)town_portal_caster_id].CanCastSpell(0x14u);
+                        pMessageQueue_50CBD0->AddGUIMessage(UIMSG_Escape, 1, 0);
+                        continue;
```

Comment:
There are the same code at the end of the case. These three lines could be removed.